### PR TITLE
Support native compilation of packages

### PR DESCRIPTION
**Pull request by flatwhatson**
This adds support for native compilation (aka. "gccemacs"), where
byte-compiled elisp is compiled to native code. Modules will be native
compiled when they are byte-compiled, unless `:no-native-compile` or
`straight-disable-native-compilation` are set.

Has no effect on emacs versions without the `native-compile` function.

Possibly addresses #492.

### CODE REVIEW

**Comment by flatwhatson** [comment]
The main challenge with native compilation is that it's *slow*.  The `comp` module uses a job queue and a pool of workers to maximize concurrency, but this is only effective if you keep the queue topped-up.  Any approach where `straight-use-package` blocks until the package is compiled won't perform well.

I've provided a "low-level" solution for this with `straight--wait-for-async-jobs`, so you can do something like the following:
```lisp
;; install the packages without blocking
(let ((straight--wait-for-async-jobs nil))
  (straight-use-package 'foo)
  (straight-use-package 'bar)
  ...)
;; block until compilation has finished
(straight--wait-for-async-jobs)
```

This could potentially be wrapped up into a nicer API:
```lisp
(straight-with-async-compilation
  (straight-use-package 'foo)
  (straight-use-package 'bar)
  ...)
```

Maybe this could even be extended to support async byte compilation?

I'm interested in any and all suggestions!

File: straight.el [hunk 1]
```diff
@@ -4252,6 +4257,61 @@ repository."
          (straight--build-dir package)
          0 'force)))))
 
+;;;;; Native-compilation
```
Comment (raxod502) [inline straight.el:4260]:
```suggestion
;;;;; Native compilation
```

File: straight.el [hunk 2]
```diff
@@ -4252,6 +4257,61 @@ repository."
...
+      (let ((inhibit-message t)
+            (message-log-max nil))
+        (native-compile-async
+         (straight--build-dir package)
+         'recursively)))
+    (straight--wait-for-async-jobs)))
+
+(defun straight--pending-async-jobs ()
+  "How many async compilation jobs are queued or in-progress.
+Returns nil if there are no pending jobs."
```
Comment (raxod502) [inline straight.el:4292]: Why does this return nil instead of zero when there are no jobs?

File: straight.el [hunk 3]
```diff
@@ -4252,6 +4257,61 @@ repository."
...
+
+(defun straight--wait-for-async-jobs ()
+  "Block until all async compilation jobs have completed (maybe).
+Blocking can be suppressed by setting `straight--wait-for-async-jobs' to nil."
+  (when (and (fboundp 'comp-async-runnings)
+             straight--wait-for-async-jobs)
+    (let ((inhibit-message t)
+          (message-log-max nil))
+      (while (not (zerop (comp-async-runnings)))
+        (sleep-for 0.1)))))
```
Comment (raxod502) [inline straight.el:4311]: Have you found this to be reliable and not too bad for performance? I have tried to write code in `straight.el` before that synchronously waits for an async job to complete, and it's been quite difficult. If this works robustly, maybe I can steal the approach for my other work too (specifically #334).

File: README.md [hunk 4]
```diff
@@ -1650,6 +1650,15 @@ customize the variable `straight-disable-byte-compilation` to effect
...
 
+By specifying a non-nil value for the `:no-native-compile` attribute
+in a package's recipe, you may inhibit native compilation. You can
+also customize the variable `straight-disable-native-compilation` to
+effect this change on all recipes which do not explicitly specify a
+`:no-native-compile` attribute. Native compilation requires
+byte-compilation, so `:no-byte-compile` and
+`straight-disable-byte-compilation` also have the effect of inhibiting
+native compilation.
+
```
Comment (raxod502) [inline README.md:1661]: I think this section of the README should include enough info to explain what gccemacs is, how people get it, and what its relation to GNU Emacs is. Otherwise, how does one tell how to use the feature?

File: straight.el (hunk 2 above)
Comment (flatwhatson) [reply straight.el:4292]: This is an attempt to make the function more ergonomic, ie.
```lisp
(if (straight--pending-async-jobs) ...)
```
vs.
```lisp
(if (not (zerop (straight--pending-async-jobs))) ...)
```
It could be considered "too cute", I'm happy to just always return the count if you'd prefer.

File: straight.el (hunk 3 above)
Comment (flatwhatson) [reply straight.el:4311]: I can't make any great claims, but it seems to do the job!

File: straight.el (hunk 1 above)
Comment (flatwhatson) [reply straight.el]: Fixed!

File: straight.el (hunk 2 above)
Comment (flatwhatson) [reply straight.el:4292]: I've changed this routine to always return a number.

File: README.md (hunk 4 above)
Comment (flatwhatson) [reply README.md:1680]: Good point, I've added some additional info, at least it should point people in the right direction.

File: straight.el [hunk 5]
```diff
@@ -4252,6 +4257,60 @@ repository."
...
+      (let ((inhibit-message t)
+            (message-log-max nil))
+        (native-compile-async
+         (straight--build-dir package)
+         'recursively 'late)))
+    (straight--wait-for-async-jobs)))
+
+(defun straight--pending-async-jobs ()
+  "How many async compilation jobs are queued or in-progress.
+Returns nil if there are no pending jobs."
```
Comment (raxod502) [inline straight.el:4292]:
```suggestion
  "How many async compilation jobs are queued or in-progress."
```

File: straight.el (hunk 2 above)
Comment (raxod502) [reply straight.el:4292]: Yeah I do like the ergonomics, but it does violate the "principle of least surprise" rather badly so I think your revised version is better.

**Comment by alphapapa** [comment]
@raxod502 @flatwhatson  Be advised that, AFAIK, Andrea is still developing the deferred compilation features.  For example, he recently added a feature that defers compilation until a library is loaded; when that happens, compilation is started, and as each unit is compiled, its native form is loaded to replace the loaded, byte-compiled form.

Having done that, it would probably be only a few lines of code to provide for the case in which:

1.  Package is installed and byte-compiled.
2.  Package's native compilation is begun.
3.  Package is loaded.
4.  Package's native compilation is finished (this step would be per-unit, i.e. per Elisp file).

In other words, it's probably not necessary to do this in Straight.  If it currently is, it would probably be better to talk to Andrea about supporting this in `comp.el` (or I might be able to help with that, as I contributed some small changes to it).  And it may be inadvisable to try to manage the queue in Straight, especially since these features are still under development (i.e. this code might break as the native-comp branch is developed).

**Comment by flatwhatson** [comment]
Understood, I think it's clear that this is very much "experimental" and will probably need review as the feature development progresses.

The deferred compilation feature is very nice, and you're right that straight needs to do very little (if anything) when used that way.  The functionality I've added here is for a different use-case; pre-compiling packages so they're *not* building in the background.

**Comment by alphapapa** [comment]
> The functionality I've added here is for a different use-case; pre-compiling packages so they're _not_ building in the background.

That's what I'm talking about.  The difference between starting compilation at package-load time and starting compilation at package-install time is negligible.  Either way, we want the files to be compiled in a non-blocking way, and when each compilation unit is finished compiling, if its byte-compiled form is already loaded, we want its native-compiled form to be loaded instead.  Support for that should be done in `comp.el` and `package.el`, not in Straight.

**Comment by flatwhatson** [comment]
We use the `native-compile-async` function provided by `comp.el` to request compilation of all elisp files in the package directory.  This ensures that all features provided by that package will be compiled up-front.  This approach also allows us to explicitly disable compilation of a package using its recipe.

In contrast, relying on the deferred compilation system would *not* guarantee that everything was compiled, and does *not* provide a way to blacklist compilation of a package.  Many packages are split into a number of modules and use autoloads to load only when needed, so for infrequently-used modules package-load time could be much later than package-install time.

I don't think anything has been added here which belongs upstream, but if you can point to something specific I'd be very happy to contribute!

**Comment by alphapapa** [comment]
> In contrast, relying on the deferred compilation system would _not_ guarantee that everything was compiled, and does _not_ provide a way to blacklist compilation of a package. Many packages are split into a number of modules and use autoloads to load only when needed, so for infrequently-used modules package-load time could be much later than package-install time.

There seems to be a communication problem here.  Please read my comments carefully.  I am aware of how the deferred compilation system currently works, and I am aware of how you want to compile packages at install-time rather than waiting for load-time.  My point is that a few simple changes to `comp.el` could support both use cases, which would also allow AOT compilation using other package managers like `package.el`, and it would mean that Straight would not need to maintain this code itself.

**Comment by flatwhatson** [comment]
OK, that makes more sense.  Thanks for your patience!

I believe that our use-case is already well-supported by `comp.el`, and that similar functionality could be added to `package.el` without too much effort.  I don't use `package.el`, so am probably not the right person to do this work.

In the simplest form, we're doing the following:
```lisp
(native-compile-async "/path/to/package1" 'recursively)
(native-compile-async "/path/to/package2" 'recursively)
...
(while (not (zerop (comp-async-runnings)))
  (sleep-for 0.1))
```

Perhaps `comp.el` could add a more convenient routine for blocking until complation has finished?  Otherwise I'm not sure how things could be improved from that end.

**Comment by alphapapa** [comment]
Blocking until compilation is finished seems like a very bad idea, because that can mean many minutes of waiting.  That's why it's done asynchronously, and that's the point of deferred compilation: the user can continue using Emacs, including using the new package immediately, and the native-compiled units replace the byte-compiled ones when ready.

AFAIU, the only difference in the desired behavior here is to begin compilation immediately, rather than waiting until a package is loaded.  And that can be done very easily by calling `native-compile-async`.  The remaining issue is to ensure that, if a compilation unit's byte-compiled form is loaded before the native-compiled form is ready, the native-compiled form is loaded automatically when compilation finishes.

AFAIR, this is currently handled by setting a variable or argument when compilation begins.  But since the package's files would not already be loaded when Straight asks them to be compiled, that would not happen, so if the user loaded the package's features before compilation finishes, the native-compiled forms would not be loaded automatically.

So, again, I think this would be trivial to handle cleanly in `comp.el`.  I think Straight shouldn't be doing anything more than `(native-compile-async "/path/to/package1" 'recursively)`.  And it definitely shouldn't be blocking until compilation finishes.

**Comment by alphapapa** [comment]
> AFAIR, this is currently handled by setting a variable or argument when compilation begins.

I see that you did this in https://github.com/raxod502/straight.el/pull/507/commits/130b8a24c99e8f633233083c7e58091bcc0ad880.  So:

1.  From what I can tell, doing that forces the file to load when compilation finishes, even if the user hasn't already loaded it.  That's not the end of the world, but it shouldn't be the case.  This is what should be handled by minor improvements in `comp.el`.
2.  Why would you ever want to block while waiting for compilation?  That seems to defeat the purpose of asynchronous compilation.

Since this PR is merged already, I'll open an issue.
